src/main/java/hudson/remoting/PingFailureAnalyzer.java

```diff
@@ -0,0 +1,225 @@
+     *
+     */
+    @IgnoreJRERequirement
+    @edu.umd.cs.findbugs.annotations.SuppressWarnings(
+            value = {"VA_FORMAT_STRING_USES_NEWLINE"},
+            justification = "We don't want platform specific"
+    )
+    public void takeThreadDump() {
+        long iota = System.currentTimeMillis();
+        final SimpleDateFormat format = new SimpleDateFormat("yyyyMMdd-HHmmss.SSS");
+        File file = threadDumps.file(format.format(new Date(iota++)) + ".txt");
+        threadDumps.add(file);
+
+        PrintWriter writer = null;
+        try {
+            writer = new PrintWriter(file, "UTF-8");
+        } catch (FileNotFoundException e) {
+            e.printStackTrace();
+        } catch (UnsupportedEncodingException e) {
+            e.printStackTrace();
+        }
+
+        ThreadMXBean mbean = ManagementFactory.getThreadMXBean();
+        ThreadInfo[] threads;
+        try {
+            threads = mbean.dumpAllThreads(mbean.isObjectMonitorUsageSupported(), mbean.isSynchronizerUsageSupported());
+        } catch (UnsupportedOperationException x) {
+            x.printStackTrace(writer);
+            threads = new ThreadInfo[0];
+        }
+
+        for (int ti = threads.length - 1; ti >= 0; ti--) {
+            printThreadInfo(writer, threads[ti], mbean);
+        }
+
+        // Print any information about deadlocks.
+        long[] deadLocks;
+        try {
+            deadLocks = mbean.findDeadlockedThreads();
+        } catch (UnsupportedOperationException x) {
+            x.printStackTrace(writer);
+            deadLocks = null;
+        }
+        if (deadLocks != null && deadLocks.length != 0) {
+            writer.println(" Deadlock Found ");
+            ThreadInfo[] deadLockThreads = mbean.getThreadInfo(deadLocks);
+            for (ThreadInfo threadInfo : deadLockThreads) {
+                StackTraceElement[] elements = threadInfo.getStackTrace();
+                for (StackTraceElement element : elements) {
+                    writer.println(element.toString());
+                }
+            }
+        }
+
+        writer.println();
+        writer.flush();
+    }
+
+    /**
+     * Prints the {@link ThreadInfo} (because {@link ThreadInfo#toString()} caps out the stack trace at 8 frames)
+     *
+     * @param writer the writer to print to.
+     * @param t      the thread to print
+     * @param mbean  the {@link ThreadMXBean} to use.
+     */
+    @edu.umd.cs.findbugs.annotations.SuppressWarnings(
+            value = {"VA_FORMAT_STRING_USES_NEWLINE"},
+            justification = "We don't want platform specific"
+    )
+    public static void printThreadInfo(PrintWriter writer, ThreadInfo t, ThreadMXBean mbean) {
+        long cpuPercentage;
+        try {
+            long cpuTime = mbean.getThreadCpuTime(t.getThreadId());
+            long threadUserTime = mbean.getThreadUserTime(t.getThreadId());
+            cpuPercentage = (cpuTime == 0) ? 0: 100 * threadUserTime / cpuTime;
```

Comment:
this is a bug copied from the support core plugin - please don't duplicate it :-)

100% cpu time is for things that have been running without stopping since Jenkins started.
a thread that wakes up every day and runs for 1ms should not be listed at 100%